### PR TITLE
I want to be informed about every item being out of stock

### PR DESCRIPTION
**Pull request by bartoszpietrzak1994**
```
    Scenario: Reordering previously placed order when one of items is out of stock
        Given the product "Angel T-Shirt" is out of stock
        When I browse my orders
        And I click reorder button next to the order "#00000666"
        Then I should be on my cart summary page
        And I should be notified that product "Angel T-Shirt" is out of stock
        And I should be notified that previous order total was "$59.00"
        And I should see exactly 2 notifications

    Scenario: Reordering previously placed order when several items are out of stock
        Given the product "Angel T-Shirt" is out of stock
        Given the product "Awesome Mug" is out of stock
        When I browse my orders
        And I click reorder button next to the order "#00000666"
        Then I should be on my cart summary page
        And I should be notified that products "Angel T-Shirt", "Awesome Mug" are out of stock
        And I should be notified that previous order total was "$59.00"
        And I should see exactly 2 notifications
```

Scenarios fulfilled

### CODE REVIEW

File: features/being_notified_about_changes_in_order.feature [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+Feature: Being notified about changes in order
+    In order to be aware of disabled promotions or changed order total
+    As a Customer
+    I want to be notified about such changes when I reorder previously placed order
+
+    Background:
+        Given the store operates on a single channel in the "United States" named "Web"
+        And the store has a product "Angel T-Shirt" priced at "$39.00"
+        And there are 25 units of product "Angel T-Shirt" available in the inventory
+        And this product is tracked by the inventory
```
Comment (GSadee) [inline features/being_notified_about_changes_in_order.feature]: I guess that there is no need for this step, because the product is set as tracked in the previous step

File: features/being_notified_about_changes_in_order.feature [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        And I click reorder button next to the order "#00000666"
+        Then I should be on my cart summary page
+        And I should be notified that product "Angel T-Shirt" is out of stock
+        And I should be notified that previous order total was "$59.00"
+        And I should see exactly 2 notifications
+
+    @ui
+    Scenario: Reordering previously placed order when several items are out of stock
+        Given the product "Angel T-Shirt" is out of stock
+        Given the product "Awesome Mug" is out of stock
```
Comment (GSadee) [inline features/being_notified_about_changes_in_order.feature]: `And`

File: features/reordering_previously_placed_order.feature [hunk 3]
```diff
@@ -25,38 +27,15 @@ Feature: Reordering previously placed order
         Then I should be on my cart summary page
         And I should see "Angel T-Shirt" with quantity 1 in my cart
         And my cart total should be "$19.00"
+        And I should see exactly 0 notifications
```
Comment (GSadee) [inline features/reordering_previously_placed_order.feature]: `And I should not see any notifications` or sth similar?

File: spec/ReorderEligibility/ItemsOutOfStockEligibilityCheckerSpec.php [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+use Sylius\Component\Core\Model\OrderInterface;
+use Sylius\Component\Core\Model\OrderItemInterface;
+use Sylius\Component\Core\Model\ProductVariantInterface;
+use Sylius\CustomerReorderPlugin\ReorderEligibility\ItemsOutOfStockEligibilityChecker;
+use Sylius\CustomerReorderPlugin\ReorderEligibility\ReorderEligibilityChecker;
+use Sylius\CustomerReorderPlugin\ReorderEligibility\ReorderEligibilityConstraintMessageFormatterInterface;
+
+final class ItemsOutOfStockEligibilityCheckerSpec extends ObjectBehavior
+{
+    function let(ReorderEligibilityConstraintMessageFormatterInterface $reorderEligibilityConstraintMessageFormatter)
```
Comment (GSadee) [inline spec/ReorderEligibility/ItemsOutOfStockEligibilityCheckerSpec.php]: Missing return types here and in other functions of this class

File: src/Factory/OrderFactory.php [hunk 5]
```diff
@@ -75,6 +76,13 @@ private function copyOrderItemsToReorder(OrderInterface $order, OrderInterface $
 
         /** @var OrderItemInterface $orderItem */
         foreach ($orderItems as $orderItem) {
+
```
Comment (GSadee) [inline src/Factory/OrderFactory.php]: Unnecessary blank line

File: src/ReorderEligibility/ReorderEligibilityChecker.php [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+interface ReorderEligibilityChecker
+{
+    /**
+     * Returns array of arrays
+     *
+     * @param OrderInterface $order
+     * @param OrderInterface $reorder
+     * @return mixed
+     */
+    public function check(OrderInterface $order, OrderInterface $reorder);
```
Comment (GSadee) [inline src/ReorderEligibility/ReorderEligibilityChecker.php]: Missing return type

File: src/ReorderEligibility/ReorderEligibilityChecker.php [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+
+interface ReorderEligibilityChecker
+{
+    /**
+     * Returns array of arrays
+     *
+     * @param OrderInterface $order
+     * @param OrderInterface $reorder
+     * @return mixed
+     */
```
Comment (GSadee) [inline src/ReorderEligibility/ReorderEligibilityChecker.php]: Unnecessary docblock

File: src/ReorderEligibility/ReorderEligibilityConstraintMessageFormatter.php [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+namespace Sylius\CustomerReorderPlugin\ReorderEligibility;
+
+final class ReorderEligibilityConstraintMessageFormatter implements ReorderEligibilityConstraintMessageFormatterInterface
+{
+    public function format(array $messageParameters): string
+    {
+        $message = '';
+
+        if (count($messageParameters) === 1) {
+            $message = array_pop($messageParameters);
```
Comment (GSadee) [inline src/ReorderEligibility/ReorderEligibilityConstraintMessageFormatter.php]: Early return?

File: src/Resources/translations/flashes.en.yml [hunk 9]
```diff
@@ -1,5 +1,6 @@
 sylius:
     reorder:
-        items_price_changed: 'Prices of some products has changed, which have affected order total.'
         previous_order_total: 'Previous order total: %order_total%'
-        promotion_not_enabled: 'Some promotions are no longer enabled. It may have affected order total.'
+        items_price_changed: 'Prices of products: %product_names% have changed, which have affected order total.'
+        promotion_not_enabled: 'Following promotions: %promotion_names% are no longer enabled, which have affected order total.'
+        items_out_of_stock: 'Following items: %order_items% are out of stock. It may have affected order total.'
```
Comment (GSadee) [inline src/Resources/translations/flashes.en.yml]: Wrong order of keys

File: src/Factory/OrderFactory.php [hunk 10]
```diff
@@ -75,6 +76,12 @@ private function copyOrderItemsToReorder(OrderInterface $order, OrderInterface $
 
         /** @var OrderItemInterface $orderItem */
         foreach ($orderItems as $orderItem) {
+            /** @var ProductVariantInterface $productVariant */
+            $productVariant = $orderItem->getVariant();
+            if (!($productVariant->isTracked() && $productVariant->isInStock())) {
```
Comment (GSadee) [inline src/Factory/OrderFactory.php]: IMO this if should look like: `if ($productVariant->isTracked() && !$productVariant->isInStock())`

File: src/ReorderEligibility/ItemsOutOfStockEligibilityChecker.php [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+        if (empty($variantsOutOfStock)) {
+            return [];
+        }
+
+        return [
+            'type' => 'info',
+            'message' => 'sylius.reorder.items_out_of_stock',
+            'parameters' => [
+                '%order_items%' => $this->reorderEligibilityConstraintMessageFormatter->format($variantsOutOfStock),
+            ]
```
Comment (Zales0123) [inline src/ReorderEligibility/ItemsOutOfStockEligibilityChecker.php]: I'm wondering does ReorderEligibility**Checker** should return a serialized flash message... It's against SRP principle for me. What I would expect from such a service is returning boolean flag or event better - throwing an exception if something is wrong. Then these exceptions can be caught where the main checker is used and passed to some "notifier" that would create a flash message basing on an exception.

File: src/ReorderEligibility/ItemsOutOfStockEligibilityChecker.php [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+    public function check(OrderInterface $order, OrderInterface $reorder): array
+    {
+        $variantsOutOfStock = [];
+
+        /** @var OrderItemInterface $orderItem */
+        foreach ($order->getItems()->getValues() as $orderItem) {
+            /** @var ProductVariantInterface $productVariant */
+            $productVariant = $orderItem->getVariant();
+            if ($productVariant->isTracked() && !$productVariant->isInStock()) {
+                array_push($variantsOutOfStock, $orderItem->getVariantName());
```
Comment (Zales0123) [inline src/ReorderEligibility/ItemsOutOfStockEligibilityChecker.php]: The preferred way to add something to an array is `$variantsOutOfStock[] = $orderItem->getVariantName()`

File: src/ReorderEligibility/ReorderItemPricesEligibilityChecker.php [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+
+        /** @var OrderItemInterface $reorderItem */
+        foreach ($reorder->getItems()->getValues() as $reorderItem) {
+            $reorderVariantNamesToTotal[$reorderItem->getVariantName()] = $reorderItem->getUnitPrice();
+        }
+
+        $orderVariantsWithChangedPrice = [];
+
+        foreach (array_keys($orderVariantNamesToTotal) as $variantName) {
+            if (!array_key_exists($variantName, $reorderVariantNamesToTotal)) {
```
Comment (Zales0123) [inline src/ReorderEligibility/ReorderItemPricesEligibilityChecker.php]: `!isset($variantName, $reorderVariantNamesToTotal)`

Comment (bartoszpietrzak1994) [reply src/ReorderEligibility/ReorderItemPricesEligibilityChecker.php]: I want to check if the key exists in the array, not if the value for given key is set.

File: src/ReorderEligibility/ReorderEligibilityCheckerResponse.php [hunk 14]
```diff
@@ -0,0 +1,38 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Sylius\CustomerReorderPlugin\ReorderEligibility;
+
+class ReorderEligibilityCheckerResponse
```
Comment (Zales0123) [inline src/ReorderEligibility/ReorderEligibilityCheckerResponse.php]: I think we've misunderstood about this class a little bit. What I was thinking about was:
1. Class `EligibilityCheckerResult` which contains some message what has gone wrong (and maybe a status? success/failure?)
2. Each checkers returns array/collection of `EligibilityCheckerResult`s 
3. Composite checker merges all responses into returns them to the controller
4. We need **one** more service that would convert each *result* into the flash message 

And that's all. I didn't dig deep into these `ReorderEligibilityCheckerResponseProcessor`s but I think they're just not needed. I think we should discuss it in live conversation.

File: spec/Reorder/ReordererSpec.php [hunk 15]
```diff
@@ -132,41 +131,110 @@ function it_notifies_when_promotion_is_no_longer_available(
...
+        $reorderEligibilityChecker->check($order, $reorder)->willReturn([$reorderEligibilityCheckerResponse]);
+
+        $orderFactory->createFromExistingOrder($order, $channel)->willReturn($reorder);
+        $entityManager->persist($reorder)->shouldBeCalled();
+        $entityManager->flush()->shouldBeCalled();
+
+        $this->reorder($order, $channel);
+    }
+
+    function it_checks_if_price_of_any_item_has_changed(
```
Comment (GSadee) [inline spec/Reorder/ReordererSpec.php]: Try to keep the order of parameters from `let` method here and in other method in spec files

File: spec/ReorderEligibility/ReorderItemPricesEligibilityCheckerSpec.php [hunk 16]
```diff
@@ -51,7 +52,8 @@ function it_returns_empty_array_when_prices_are_the_same(
         $secondOrderItem->getVariantName()->willReturn('test_variant_name_02');
         $secondOrderItem->getUnitPrice()->willReturn(100);
 
-        $this->check($order, $reorder)->shouldReturn([]);
+        $response = $this->check($order, $reorder);
```
Comment (GSadee) [inline spec/ReorderEligibility/ReorderItemPricesEligibilityCheckerSpec.php]: Why did you change it?

File: spec/ReorderEligibility/ReorderItemPricesEligibilityCheckerSpec.php [hunk 17]
```diff
@@ -82,12 +84,10 @@ function it_returns_violation_message_when_some_prices_are_different(
             'test_variant_name_02'
         ])->willReturn('test_variant_name_01, test_variant_name_02');
 
-        $this->check($order, $reorder)->shouldReturn([
-            'type' => 'info',
-            'message' => 'sylius.reorder.items_price_changed',
-            'parameters' => [
-                '%product_names%' => 'test_variant_name_01, test_variant_name_02'
-            ]
+        $response = $this->check($order, $reorder);
```
Comment (GSadee) [inline spec/ReorderEligibility/ReorderItemPricesEligibilityCheckerSpec.php]: As above

**Comment by GSadee** [comment]
Thanks Bartek 🤘
